presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java

```diff
@@ -0,0 +1,228 @@
+
+        extractFromSource(hit);
+        if (hit.getSourceRef() != null) {
+            totalBytes += hit.getSourceRef().length();
+        }
+        return true;
+    }
+
+    @Override
+    public boolean getBoolean(int field)
+    {
+        checkFieldType(field, ImmutableSet.of(BOOLEAN));
+        return (Boolean) getFieldValue(field);
+    }
+
+    @Override
+    public long getLong(int field)
+    {
+        checkFieldType(field, ImmutableSet.of(BIGINT, INTEGER));
+        return (Integer) getFieldValue(field);
+    }
+
+    @Override
+    public double getDouble(int field)
+    {
+        checkFieldType(field, ImmutableSet.of(DOUBLE));
+        return (Double) getFieldValue(field);
+    }
+
+    @Override
+    public Slice getSlice(int field)
+    {
+        checkFieldType(field, ImmutableSet.of(VARCHAR));
+
+        Object value = getFieldValue(field);
+        if (value instanceof Collection) {
+            return utf8Slice(VALUE_CODEC.toJson(value));
+        }
+        if (value == null) {
+            return EMPTY_SLICE;
+        }
+        return utf8Slice(String.valueOf(value));
+    }
+
+    @Override
+    public Object getObject(int field)
+    {
+        return serializeObject(columnHandles.get(field).getColumnType(), null, getFieldValue(field));
+    }
+
+    @Override
+    public boolean isNull(int field)
+    {
+        checkArgument(field < columnHandles.size(), "Invalid field index");
+        return getFieldValue(field) == null;
+    }
+
+    private void checkFieldType(int field, Set<Type> expectedTypes)
+    {
+        checkArgument(expectedTypes.contains(getType(field)), "Field %s expected type inconsistent with %s", field, getType(field));
+    }
+
+    @Override
+    public void close()
+    {
+        builder.close();
+    }
+
+    private List<SearchHit> sendElasticQuery(ElasticsearchQueryBuilder queryBuilder)
+    {
+        ImmutableList.Builder<SearchHit> result = ImmutableList.builder();
+        SearchResponse response = queryBuilder.buildScrollSearchRequest().execute().actionGet(timeout.toMillis());
+
+        if (response.getHits().getTotalHits() > maxHits) {
+            throw new PrestoException(ELASTICSEARCH_MAX_HITS_EXCEEDED,
+                    format("The number of hits for the query (%d) exceeds the configured max hits (%d)", response.getHits().getTotalHits(), maxHits));
+        }
+
+        while (true) {
+            for (SearchHit hit : response.getHits().getHits()) {
+                result.add(hit);
+            }
+            response = queryBuilder.prepareSearchScroll(response.getScrollId())
+                    .execute()
+                    .actionGet(timeout.toMillis());
+            if (response.getHits().getHits().length == 0) {
+                break;
+            }
+        }
+        return result.build();
+    }
+
+    private void setFieldIfExists(String jsonPath, Object jsonValue)
+    {
+        if (jsonPathToIndex.containsKey(jsonPath)) {
```

Comment:
What will happen if the field doesn't exist?

Reply:
if the field does not exist, it is a Null value. will set nothing in fields.